src/protocol/sort/reveal.rs

```diff
@@ -0,0 +1,107 @@
+use crate::{
+    error::BoxError,
+    field::Field,
+    helpers::{
+        mesh::{Gateway, Mesh},
+        prss::SpaceIndex,
+        Direction,
+    },
+    protocol::{context::ProtocolContext, RecordId, Step},
+    secret_sharing::Replicated,
+};
+use embed_doc_image::embed_doc_image;
+use serde::{Deserialize, Serialize};
+
+/// A message sent by each helper when they've revealed their own shares
+#[derive(Debug, Serialize, Deserialize, Default)]
+pub struct AValue<F> {
+    a: F,
+}
+
+/// This implements reveal algorithm of "Efficient Secure Three-Party Sorting Protocol with an Honest Majority" at communication cost of 3R.
+/// Revealing is the protocol that on input a pair of shares, outputs a secret. For simplicity, we
+/// consider a simple revealing in which each Pi sends [a]i to Pi+1 and then reconstructs a from [a]i and
+/// [a]i−1.
+// Input: Each helpers know their own secret shares
+// Output: At the end of the protocol, all 3 helpers know a revealed (or opened) secret
+#[derive(Debug)]
+pub struct Reveal<F> {
```

Comment:
I wonder if we should make this protocol sort (or even apply/applyinv) specific. i.e. impossible to apply outside of the allowed boundaries of sort. I worry that someone may instruct helpers to reveal their secret shares they've received from report collectors using this (because it takes just a replicated secret share). Maybe wrapping a replicated secret share into its own type is sufficient, so taking `PermutedShare<F>` instead of `Replicated<F>` may be enough.

this is just an idea, curious if we should discuss it further or if the reasoning here makes sense

Reply:
I don't think we need to worry about this. The helpers will not be doing anything but running this Rust crate. They aren't writing any code. If we call this at the wrong time, we are in trouble. I don't think we need to try to hide this function from ourselves.

The helpers don't just respond to commands right? That would be a problem.

Reply:
artificially making our life harder by trying to hide it would not help at all, I agree. I maybe overindexing on this because I've seen cases where something that wasn't intended to be used on real data actually sneaked into production code. Because it looked legit at the review phase,  this bug did not manifest itself until it was too late. 

While making this `Reveal` input "incompatible" with the input for regular `Reveal` function that would be used in tests does not guarantee correctness on its own, it may provide some defense in depth, because it won't be easily pluggable into the pipeline of transforming secret shares. Compiler will actually reject this. Of course someone may still try to use it, but hopefully it will trigger a conversation at the PR stage.

Reply:
actually, it is probably enough to make sure 'Reveal' is not accessible from outside the sort module...

Reply:
I tried to include a TestReveal under cfg(test) - based on definition, would this code compile only during test run?

Reply:
> actually, it is probably enough to make sure 'Reveal' is not accessible from outside the sort module...

But we need a "reveal" for other things as well, such as the CheckZero functionality. I'd like to avoid duplicating that code.

Reply:
Oh ya I totally forgot about that being used outside sort under malicious setting. Let me revert this change and bring it back to original form

Reply:
hmm, that makes me feel that we may want to create `reveal` method on `ProtocolContext` that returns `Reveal` similar to what currently `multiply` does. We would want `Reveal` constructors to be hidden from IPA code (using pub(super) or pub(in path) https://doc.rust-lang.org/reference/visibility-and-privacy.html#pubin-path-pubcrate-pubsuper-and-pubself). The idea is that only `ProtocolContext` should be able to instantiate it and we can verify that the call site is allowed to reveal by checking the step variable.

I don't see a better way to check the step rather than downcast, it is ugly but it should work.

Reply:
Made these changes. Makes sense to keep reveal in context since same as multiply, this is not scoped to only sort but will be used in other part of protocols as well. (e.g. when we implement malicious setting and need CheckZero protocol).